pkg/api/controller.go

```diff
@@ -99,6 +99,225 @@ type Controller struct {
+	storageConfig := c.getStorageConfig()
+	if !swag.BoolValue(storageConfig.PreSignMultipartUpload) {
+		writeError(w, r, http.StatusNotImplemented, "presign multipart upload API is not supported")
+		return
+	}
+
+	// validation checks
+	if uploadID == "" {
+		writeError(w, r, http.StatusBadRequest, "upload_id is required")
+		return
+	}
+
+	// verify physical address
+	if err := c.Catalog.VerifyLinkAddress(ctx, repository, body.PhysicalAddress); c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	repo, err := c.Catalog.GetRepository(ctx, repository)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	if err := c.BlockAdapter.AbortMultiPartUpload(ctx, block.ObjectPointer{
+		StorageNamespace: repo.StorageNamespace,
+		IdentifierType:   block.IdentifierTypeRelative,
+		Identifier:       body.PhysicalAddress,
+	}, uploadID); c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+	writeResponse(w, r, http.StatusNoContent, nil)
+}
+
+func (c *Controller) CompletePresignMultipartUpload(w http.ResponseWriter, r *http.Request, body apigen.CompletePresignMultipartUploadJSONRequestBody, repository string, branch string, uploadID string, params apigen.CompletePresignMultipartUploadParams) {
+	if !c.authorize(w, r, permissions.Node{
+		Permission: permissions.Permission{
+			Action:   permissions.WriteObjectAction,
+			Resource: permissions.ObjectArn(repository, params.Path),
+		},
+	}) {
+		return
+	}
+	ctx := r.Context()
+	c.LogAction(ctx, "complete_presign_multipart_upload", r, repository, branch, "")
+
+	// check if api is supported
+	storageConfig := c.getStorageConfig()
+	if !swag.BoolValue(storageConfig.PreSignMultipartUpload) {
+		writeError(w, r, http.StatusNotImplemented, "presign multipart upload API is not supported")
+		return
+	}
+
+	// verify physical address
+	repo, err := c.Catalog.GetRepository(ctx, repository)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	writeTime := time.Now()
+	physicalAddress, addressType := normalizePhysicalAddress(repo.StorageNamespace, body.PhysicalAddress)
+	if addressType != catalog.AddressTypeRelative {
+		writeError(w, r, http.StatusBadRequest, "physical address must be relative to the storage namespace")
+		return
+	}
+
+	//  verify it has been saved for linking
+	if err := c.Catalog.VerifyLinkAddress(ctx, repository, physicalAddress); c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	var multipartList []block.MultipartPart
+	for _, part := range body.Parts {
+		multipartList = append(multipartList, block.MultipartPart{
+			PartNumber: part.PartNumber,
+			ETag:       part.Etag,
+		})
+	}
+
+	mpuResp, err := c.BlockAdapter.CompleteMultiPartUpload(ctx, block.ObjectPointer{
+		StorageNamespace: repo.StorageNamespace,
+		IdentifierType:   block.IdentifierTypeRelative,
+		Identifier:       physicalAddress,
+	}, uploadID, &block.MultipartUploadCompletion{
+		Part: multipartList,
+	})
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	checksum := strings.Trim(mpuResp.ETag, `"`)
```

Comment:
Check the trim function in `LinkPhysicalAddress`, I think we should align with one implementation

Reply:
In the case of link physical address the etag comes from header to the client which calls the API.
But I'll align the code.